File: tasks/healthchecks-telemetry/stacks/go/base/code/healthchecks_telemetry_service.tpl.go
// File: healthchecks_telemetry_service.tpl.go
// Purpose: Template for healthchecks-telemetry implementation
// Generated for: {{PROJECT_NAME}}

// Healthcheckstelemetry Service for Go
// Generated for {{PROJECT_NAME}}
package healthchecks_telemetry

import (
    "context"
)

// HealthchecksTelemetryService handles healthchecks-telemetry operations
type HealthchecksTelemetryService struct {{
    config  map[string]interface{{}}
    enabled bool
    timeout int
}}

// Config holds service configuration
type Config struct {{
    Enabled bool
    Timeout int
}}

// NewHealthchecksTelemetryService creates a new service instance
func NewHealthchecksTelemetryService(cfg Config) *HealthchecksTelemetryService {{
    return &HealthchecksTelemetryService{{
        config:  make(map[string]interface{{}}),
        enabled: cfg.Enabled,
        timeout: cfg.Timeout,
    }}
}}

// Execute runs the healthchecks-telemetry service
func (s *HealthchecksTelemetryService) Execute(ctx context.Context, input map[string]interface{{}}) (map[string]interface{{}}, error) {{
    // TODO: Implement healthchecks-telemetry logic
    return map[string]interface{{}}{{"status": "success", "data": input}}, nil
}}

// GetStatus returns service health status
func (s *HealthchecksTelemetryService) GetStatus() map[string]interface{{}} {{
    return map[string]interface{{}}{{"status": "healthy", "service": "{{PROJECT_NAME}}-healthchecks-telemetry", "enabled": s.enabled, "stack": "go"}}
}}

// Shutdown gracefully stops the service
func (s *HealthchecksTelemetryService) Shutdown() error {{
    s.enabled = false
    return nil
}}

File: tasks/healthchecks-telemetry/stacks/node/base/code/healthchecks_telemetry_service.tpl.js
/**
 * Template: healthchecks_telemetry_service.tpl.js
 * Purpose: healthchecks_telemetry_service template
 * Stack: node
 * Tier: base
 */

#!/usr/bin/env node
/**
 * Template: healthchecks_telemetry_service.tpl.js
 * Purpose: Health checks and telemetry service for Node.js applications
 * Stack: node
 * Generated for: {{PROJECT_NAME}}
 */

const { EventEmitter } = require('events');

const DEFAULT_CONFIG = {
    enabled: true,
    timeout: 30000,
    maxRetries: 3,
    checkInterval: 60000,
    metricsInterval: 10000
};

class HealthchecksTelemetryService extends EventEmitter {
    constructor(config = {}) {
        super();
        this.config = { ...DEFAULT_CONFIG, ...config };
        this.enabled = this.config.enabled;
        this.timeout = this.config.timeout;
        this.initialized = false;
        this.checks = new Map();
        this.metrics = {};
    }

    async initialize() {
        if (this.initialized) return;
        try {
            // TODO: Setup health check and monitoring infrastructure
            this._startChecks();
            this.initialized = true;
            this.emit('initialized');
        } catch (error) {
            this.emit('error', error);
            throw error;
        }
    }

    _startChecks() {
        setInterval(() => this._runHealthChecks(), this.config.checkInterval);
        setInterval(() => this._collectMetrics(), this.config.metricsInterval);
    }

    async _runHealthChecks() {
        const results = {};
        for (const [name, check] of this.checks) {
            try {
                results[name] = await check();
            } catch (error) {
                results[name] = { status: 'failed', error: error.message };
            }
        }
        this.emit('healthCheck', results);
    }

    async _collectMetrics() {
        this.metrics = {
            timestamp: Date.now(),
            memory: process.memoryUsage(),
            uptime: process.uptime(),
            cpuUsage: process.cpuUsage()
        };
        this.emit('metrics', this.metrics);
    }

    async execute(inputData) {
        if (!this.enabled) return { status: 'disabled', message: 'Service is disabled' };

        const startTime = Date.now();
        try {
            // TODO: Implement healthchecks-telemetry logic here
            const result = await this._processRequest(inputData);
            const responseTime = Date.now() - startTime;
            this.emit('success', { responseTime });
            return { status: 'success', data: result, responseTime };
        } catch (error) {
            const responseTime = Date.now() - startTime;
            this.emit('error', { error, responseTime });
            return { status: 'error', error: error.message, responseTime };
        }
    }

    async _processRequest(requestData) {
        const { action } = requestData;
        switch (action) {
            case 'health': return this._getHealth();
            case 'metrics': return this._getMetrics();
            case 'register': return this._registerCheck(requestData);
            default: throw new Error(`Unknown action: ${action}`);
        }
    }

    async _getHealth() {
        return {
            status: 'healthy',
            timestamp: Date.now(),
            checks: Object.fromEntries(this.checks)
        };
    }

    async _getMetrics() {
        return this.metrics;
    }

    async _registerCheck({ name, check }) {
        this.checks.set(name, check);
        return { registered: name };
    }

    async validate(inputData) {
        return inputData && typeof inputData === 'object' && inputData.action;
    }

    getStatus() {
        return {
            status: this.initialized ? 'healthy' : 'not_initialized',
            service: '{{PROJECT_NAME}}-healthchecks-telemetry',
            enabled: this.enabled,
            stack: 'node',
            checks: this.checks.size,
            uptime: process.uptime()
        };
    }

    async shutdown() {
        try {
            this.checks.clear();
            this.initialized = false;
            this.emit('shutdown');
        } catch (error) {
            this.emit('error', error);
            throw error;
        }
    }
}

module.exports = { HealthchecksTelemetryService, DEFAULT_CONFIG };
